File: go/terminal/text/wrap_test.go
package text

import (
	"reflect"
	"testing"
)

// Fixed-advance measure: 7px per rune. Wrapping is pure geometry over it.
func meas(s string) float32 { return float32(len([]rune(s))) * 7 }

func TestWrapSingleLineFits(t *testing.T) {
	got := Wrap(meas, "ab cd", 100)
	want := []Line{{0, 5}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapBreaksAtWordBoundary(t *testing.T) {
	// 19 runes; 14 fit (98px <= 100). The fitted prefix ends at a word
	// end, so the break happens there.
	got := Wrap(meas, "aaaa aaaa aaaa aaaa", 100)
	want := []Line{{0, 14}, {15, 19}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapEatsBreakSpaces(t *testing.T) {
	// The spaces consumed at a break fall in the gap between lines, so
	// s[l.Start:l.End] never carries them.
	got := Wrap(meas, "aa   bb", 30) // 4 runes fit (28px)
	want := []Line{{0, 2}, {5, 7}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapHardBreaksOverlongWord(t *testing.T) {
	got := Wrap(meas, "aaaaaaaaaaaaaaa", 70) // 15 runes, 10 per line
	want := []Line{{0, 10}, {10, 15}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapHonorsNewlines(t *testing.T) {
	got := Wrap(meas, "ab\n\ncd", 100)
	want := []Line{{0, 2}, {3, 3}, {4, 6}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapTrailingNewlineYieldsEmptyLine(t *testing.T) {
	// A textarea caret can sit on the empty line after a trailing \n.
	got := Wrap(meas, "ab\n", 100)
	want := []Line{{0, 2}, {3, 3}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapEmptyString(t *testing.T) {
	got := Wrap(meas, "", 100)
	want := []Line{{0, 0}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapMultibyteOffsetsAreBytes(t *testing.T) {
	// "éé éé" - é is 2 bytes; offsets must be byte positions.
	got := Wrap(meas, "éé éé", 30) // 4 runes fit; break at the space
	want := []Line{{0, 4}, {5, 9}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

func TestWrapAlwaysAdvances(t *testing.T) {
	// Pathologically narrow: still one rune per line, never an infinite loop.
	got := Wrap(meas, "abc", 1)
	want := []Line{{0, 1}, {1, 2}, {2, 3}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %v want %v", got, want)
	}
}

File: src/test/wrap_test.ts
import { wrapText } from '../gfx/text/wrap';
import { eq, test } from './harness';

// Fixed-advance measure: 7px per code point. These fixtures mirror
// go/terminal/text/wrap_test.go
const meas = (s: string): number => [...s].length * 7;

test('single line fits', () => {
  eq(wrapText(meas, 'ab cd', 100), [{ start: 0, end: 5 }]);
});

test('breaks at word boundary', () => {
  // 19 units; 14 fit. the fitted prefix ends at a word end
  eq(wrapText(meas, 'aaaa aaaa aaaa aaaa', 100), [
    { start: 0, end: 14 },
    { start: 15, end: 19 },
  ]);
});

test('break spaces fall in the gap', () => {
  eq(wrapText(meas, 'aa   bb', 30), [
    { start: 0, end: 2 },
    { start: 5, end: 7 },
  ]);
});

test('overlong word hard-breaks', () => {
  eq(wrapText(meas, 'aaaaaaaaaaaaaaa', 70), [
    { start: 0, end: 10 },
    { start: 10, end: 15 },
  ]);
});

test('honors newlines', () => {
  eq(wrapText(meas, 'ab\n\ncd', 100), [
    { start: 0, end: 2 },
    { start: 3, end: 3 },
    { start: 4, end: 6 },
  ]);
});

test('trailing newline yields an empty line', () => {
  eq(wrapText(meas, 'ab\n', 100), [
    { start: 0, end: 2 },
    { start: 3, end: 3 },
  ]);
});

test('empty string is one empty line', () => {
  eq(wrapText(meas, '', 100), [{ start: 0, end: 0 }]);
});

test('never splits a surrogate pair', () => {
  // each emoji is 2 UTF-16 units; two fit per line (14px ≤ 20). a break at
  // unit 3 would split the second emoji so it must land at 4.
  eq(wrapText(meas, '😀😀😀', 20), [
    { start: 0, end: 4 },
    { start: 4, end: 6 },
  ]);
});

test('always advances at pathological widths', () => {
  eq(wrapText(meas, 'abc', 1), [
    { start: 0, end: 1 },
    { start: 1, end: 2 },
    { start: 2, end: 3 },
  ]);
});
